Stratis.Bitcoin/Features/BlockStore/LoopSteps/BlockStoreInnerStepFindBlocks.cs

```diff
@@ -39,43 +33,44 @@ public BlockStoreInnerStepFindBlocks(ILoggerFactory loggerFactory)
         public override async Task<InnerStepResult> ExecuteAsync(BlockStoreInnerStepContext context)
         {
             this.logger.LogTrace("()");
-            context.GetNextBlock();
-
-            if (await ShouldStopFindingBlocks(context))
-            {
-                if (!context.DownloadStack.Any())
-                {
-                    this.logger.LogTrace("(-):{0}", InnerStepResult.Stop);
-                    return InnerStepResult.Stop;
-                }
 
-                context.StopFindingBlocks();
-            }
-            else
+            while (await ShouldStopFindingBlocks(context) == false)
             {
-                context.BlockStoreLoop.BlockPuller.AskBlock(context.NextChainedBlock);
                 context.DownloadStack.Enqueue(context.NextChainedBlock);
-
-                if (context.DownloadStack.Count == context.BlockStoreLoop.BatchDownloadSize)
-                    context.StopFindingBlocks();
+                context.GetNextBlock();
             }
 
+            context.StopFindingBlocks();
+
+            if (context.DownloadStack.Any())
+                context.BlockStoreLoop.BlockPuller.AskForMultipleBlocks(context.DownloadStack.ToArray());
+
             this.logger.LogTrace("(-):{0}", InnerStepResult.Next);
+
             return InnerStepResult.Next;
         }
 
         private async Task<bool> ShouldStopFindingBlocks(BlockStoreInnerStepContext context)
         {
-            this.logger.LogTrace("()");
+            if (context.NextChainedBlock == null)
+                return true;
+
+            if (context.InputChainedBlock != null && (context.NextChainedBlock.Header.HashPrevBlock != context.InputChainedBlock.HashBlock))
+                return true;
+
+            if (context.NextChainedBlock.Height > context.BlockStoreLoop.ChainState.HighestValidatedPoW?.Height)
+                return true;
+
+            if (context.BlockStoreLoop.PendingStorage.ContainsKey(context.NextChainedBlock.HashBlock))
+                return true;
+
+            if (await context.BlockStoreLoop.BlockRepository.ExistAsync(context.NextChainedBlock.HashBlock))
+                return true;
 
-            bool res = (context.NextChainedBlock == null)
-                || (context.NextChainedBlock.Header.HashPrevBlock != context.InputChainedBlock.HashBlock)
-                || (context.NextChainedBlock.Height > context.BlockStoreLoop.ChainState.HighestValidatedPoW?.Height)
-                || (context.BlockStoreLoop.PendingStorage.ContainsKey(context.NextChainedBlock.HashBlock))
-                || (await context.BlockStoreLoop.BlockRepository.ExistAsync(context.NextChainedBlock.HashBlock));
+            if (context.DownloadStack.Count >= BlockStoreInnerStepContext.DownloadStackThreshold)
+                return true;
```

Comment:
i think it would be beneficial to put this condition first or second because its evaluation costs nothing compared to possibly other checks, especially BS.ExistAsync could probably be heavy (heavier)